src/PluginManager/Microsoft.Performance.Toolkit.PluginManager.Core/Discovery/DiscoveryManager.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.Performance.SDK;
+
+namespace Microsoft.Performance.Toolkit.PluginManager.Core.Discovery
+{
+    /// <summary>
+    ///     Manages plugin discovery related resources.
+    /// </summary>
+    // TODO: Lazy initialize discoverers
+    public class DiscoveryManager
+    {
+        private readonly ISet<IPluginSource> pluginSources;
+        private readonly HashSet<IPluginDiscovererProvider> discovererProviders;
+        private readonly IDiscovererProvidersLoader loader;
+        private readonly Dictionary<IPluginSource, IList<IPluginDiscoverer>> sourceToDiscoverers;
+
+        public DiscoveryManager(
+            IEnumerable<IPluginDiscovererProvider> discovererProviders,
+            IDiscovererProvidersLoader discovererSourceLoader)
+        {
+            this.pluginSources = new HashSet<IPluginSource>();
+            this.loader = discovererSourceLoader;
+            this.discovererProviders = new HashSet<IPluginDiscovererProvider>(discovererProviders);
```

Comment:
Do we want to use thread safe libraries?

Reply:
Addressed in the new design